executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,684 @@
+				break
+			}
+			task.innerIter = chunk.NewIterator4Chunk(task.innerResult.GetChunk(task.innerCursor))
+			task.innerIter.Begin()
+		}
+	}
+	task.sameKeyIter = chunk.NewIterator4Slice(task.sameKeyRows)
+	task.sameKeyIter.Begin()
+
+	return errors.Trace(err)
+}
+
+func (imw *innerMergeWorker) compare(outerRow, innerRow chunk.Row) (int, error) {
+	for i := 0; i < len(imw.outerMergeCtx.joinKeys); i++ {
+		cmp, _, err := imw.innerMergeCtx.compareFuncs[i](imw.ctx, imw.outerMergeCtx.joinKeys[i], imw.innerMergeCtx.joinKeys[i], outerRow, innerRow)
+		if err != nil {
+			return 0, err
+		}
+
+		if cmp != 0 {
+			return int(cmp), nil
+		}
+	}
+	return 0, nil
+}
+
+func (imw *innerMergeWorker) constructDatumLookupKeys(task *lookUpMergeJoinTask) ([][]types.Datum, error) {
+	dLookUpKeys := make([][]types.Datum, 0, task.outerResult.NumRows())
+	for i := 0; i < task.outerResult.NumRows(); i++ {
+		dLookUpKey, err := imw.constructDatumLookupKey(task, i)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if dLookUpKey == nil {
+			continue
+		}
+		dLookUpKeys = append(dLookUpKeys, dLookUpKey)
+	}
+
+	return dLookUpKeys, nil
+}
+
+func (imw *innerMergeWorker) constructDatumLookupKey(task *lookUpMergeJoinTask, rowIdx int) ([]types.Datum, error) {
+	if task.outerMatch != nil && !task.outerMatch[rowIdx] {
+		return nil, nil
+	}
+	outerRow := task.outerResult.GetRow(rowIdx)
+	sc := imw.ctx.GetSessionVars().StmtCtx
+	keyLen := len(imw.keyCols)
+	dLookupKey := make([]types.Datum, 0, keyLen)
+	for i, keyCol := range imw.outerMergeCtx.keyCols {
+		outerValue := outerRow.GetDatum(keyCol, imw.outerMergeCtx.rowTypes[keyCol])
+		// Join-on-condition can be promised to be equal-condition in
+		// IndexNestedLoopJoin, thus the filter will always be false if
+		// outerValue is null, and we don't need to lookup it.
+		if outerValue.IsNull() {
+			return nil, nil
+		}
+		innerColType := imw.rowTypes[imw.keyCols[i]]
+		innerValue, err := outerValue.ConvertTo(sc, innerColType)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		cmp, err := outerValue.CompareDatum(sc, &innerValue)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if cmp != 0 {
+			// If the converted outerValue is not equal to the origin outerValue, we don't need to lookup it.
+			return nil, nil
+		}
+		dLookupKey = append(dLookupKey, innerValue)
+	}
+	return dLookupKey, nil
+}
+
+func (imw *innerMergeWorker) dedupDatumLookUpKeys(dLookUpKeys [][]types.Datum) [][]types.Datum {
+	if len(dLookUpKeys) < 2 {
+		return dLookUpKeys
+	}
+	sc := imw.ctx.GetSessionVars().StmtCtx
+	deDupedLookupKeys := dLookUpKeys[:1]
+	for i := 1; i < len(dLookUpKeys); i++ {
+		cmp := compareRow(sc, dLookUpKeys[i], dLookUpKeys[i-1])
+		if cmp != 0 {
+			deDupedLookupKeys = append(deDupedLookupKeys, dLookUpKeys[i])
+		}
+	}
+	return deDupedLookupKeys
+}
+
+func (imw *innerMergeWorker) fetchInnerResults(ctx context.Context, task *lookUpMergeJoinTask, dLookUpKeys [][]types.Datum) error {
```

Comment:
Feels like there are some functions that are duplicated with `index join`. Can we extract a struct so we can reuse the function?

Reply:
Only a few. It's unnecessary.